packages/myst-common/src/utils.ts

```diff
@@ -179,27 +179,32 @@ export function toText(content?: Node[] | Node | null): string {
 export function copyNode<T extends Node | Node[]>(node: T): T {
   return structuredClone(node);
 }
+function nodeHasChildren(node: Content): node is Extract<Content, Parent> {
+  return 'children' in node;
+}
 
-export function mergeTextNodes(node: GenericNode): GenericNode {
-  const children = node.children?.reduce((c, n) => {
-    if (n?.type !== 'text') {
-      c.push(mergeTextNodes(n));
-      return c;
-    }
-    const last = c[c.length - 1];
-    if (last?.type !== 'text') {
-      c.push(n);
-      return c;
-    }
-    if (n.position?.end) {
-      if (!last.position) last.position = {} as Required<GenericNode>['position'];
-      last.position.end = n.position.end;
-    }
-    if (!last.value) last.value = '';
-    if (n.value) last.value += n.value;
-    return c;
-  }, [] as GenericNode[]);
-  if (children) node.children = children;
+export function mergeTextNodes(node: Content): Content {
+  const children = nodeHasChildren(node)
+    ? node.children.reduce((c, n) => {
+        if (n?.type !== 'text') {
+          c.push(mergeTextNodes(n));
+          return c;
+        }
+        const last = c[c.length - 1];
+        if (last?.type !== 'text') {
+          c.push(n);
+          return c;
+        }
+        if (n.position?.end) {
+          if (!last.position) last.position = {} as Required<Node>['position'];
+          last.position.end = n.position.end;
+        }
+        if (!last.value) last.value = '';
+        if (n.value) last.value += n.value;
+        return c;
+      }, [] as Content[])
+    : [];
+  if (children) (node as Parent).children = children;
   return node;
```

Comment:
I suspect this should actually be generic, e.g. 
```typescript
export function mergeTextNodes<T extends Nodes>(node: T): T {
```
although that won't be a lower bound